p4-16/psa/examples/psa-example-clone-to-port.p4

```diff
@@ -135,26 +136,50 @@ control egress(inout headers hdr,
 
 control computeChecksum(inout headers hdr, inout metadata meta) {}
 
-control DeparserImpl(packet_out packet, clone cl, in headers hdr) {
-    // Alternatively, we can provide a constructor to Clone().
-    // Clone() clone;
+
+control DeparserImpl(packet_out packet, inout headers hdr) {
     apply {
-        cl.emit({hdr.ethernet.srcAddr, hdr.ipv4.srcAddr});
-        // Use a different method name (pack?)
-        // cl.pack({hdr.ethernet.srcAddr, hdr.ipv4.srcAddr});
+        packet.emit(hdr.eth);
+        packet.emit(hdr.ipv4);
+    }
+}
 
-        // XXX(hanw): multiple clone instances?
+control IngressDeparserImpl(packet_out packet,
+    clone_out clone,
+    inout headers hdr,
+    in userMetadata meta,
+    in psa_ingress_output_metadata_t istd) {
+    DeparserImpl() common_deparser;
+    apply {
+        if (istd.clone_instance == 0 && istd.clone) {
+            clone.add_metadata({hdr.eth.srcAddr, hdr.eth.dstAddr});
```

Comment:
This is wonderful, but where is this consumed?

Reply:
Yes, same question from me.  An example of adding metadata should show it being used somewhere.

(1) Is it intended to be parsed explicitly in the user's P4 program?  If so, do we use the 'extract' method for that purpose?  What order does this metadata occur in relative to other contents of the packet?  Always before the rest of the packet and any headers it might contain?  If there is more than one collection of metadata that can be added to a clone, how does the parser know which case to parse?

(2) Or is it intended to be more like bridged metadata, where the values you specify in add_metadata are auto-magically copied by the implementation to the fields with the same names in the egress code?

If so, do the names and types of the fields have to match exactly between ingress and egress?

Should the compiler give an error if a field name is given to the clone.add_metadata() method that does not exist in egress metadata?  If the name does exist, but the types are different?

Your example gives fields to clone.add_metadata in the hdr struct.  Which fields should a programmer be allowed to give as arguments here?  Inside user-defined metadata seems like an obvious choice to permit, but you are also suggesting user-defined headers, too.

Reply:
If the answer to the question "Is it intended that fields given to clone.add_metadata() are 'auto-magically' copied from ingress to egress?" is _yes_, then isn't bridged metadata pretty much like saying "normal_packet.add_metadata(user_meta);", where user_meta is the struct that contains all user-defined metadata?  (and the compiler is allowed to optimize out any fields from either list of arguments, if it can tell that the field is not referenced in the egress code)

If so, such a method for normal packets seems like a way to make metadata added to clones, and opt-in bridged metadata for normal packets, be specified in a common way in P4 programs.

Reply:
I have updated the example to show the parser implementation.

The user program is responsible to parse the cloned header using the extract method. The clone header is prepended before other contents of the packet. The user can add arbitrary amount of metadata to the cloned header (up to the limit that a target supports). It is up to the user to parse the headers properly. As we only define clone from ingress to egress, and egress to egress, the parsing should be done in the egress parser. In the future, if we decided to define ingress to ingress and egress to ingress clone, the same rule applies to ingress parser. The name and types of the fields do not need to match between the deparser and parser (although it might be a good practice). But the length of the extracted header must be the same. I don't think compiler need to give an error if user decides to 'cast' the clone header to a different type upon receiving the header.